### ISSUE TITLE
Actuator (mappings) JSON for functional RMs

### ISSUE THREAD

**Issue by BoykoAlex** [spring-projects/spring-boot#15230]
STS team attempts to find a spot in the source code that declares RM using functional style using the JSON data from the Actuator.

Source code example:
```
public class VetsRouter {

	@Bean
	public RouterFunction<ServerResponse> route(VetsHandler vetsHandler) {

		return RouterFunctions
			.route(RequestPredicates.GET("/vets").and(RequestPredicates.accept(MediaType.TEXT_PLAIN)), vetsHandler::vets)
			.andRoute(RequestPredicates.GET("/vets.html"), vetsHandler::vetsHtml);
	}
	
	@Bean
	public RouterFunction<ServerResponse> moreRoute(VetsHandler vetsHandler) {

		return RouterFunctions
			.route(RequestPredicates.GET("/new-vets").and(RequestPredicates.accept(MediaType.TEXT_PLAIN)), vetsHandler::vets)
			.andRoute(RequestPredicates.GET("/new-vets.html"), vetsHandler::vetsHtml);
	}
}
```

The JSON data from the Actuator for `/vets` route:
```
          {
            "predicate": "((GET && /vets) && Accept: [text/plain])",
            "handler": "org.springframework.samples.petclinic.deviations.VetsRouter$$Lambda$868/443912630@369564bd",
            "details": {
              "handlerFunction": {
                "className": "org.springframework.samples.petclinic.deviations.VetsRouter$$Lambda$868/443912630"
              }
            }
          },
```
There are two pieces of data to help matching are __predicate__ and __handler__
The handler data should be most helpful in matching the proper place in the source. The matching place in the source code we think should be the `route` method declaration. Based on the handler's data we cannot match the `route` method unfortunately. The lambda with magic numbers is useless in this case.

It'd be great if the `details` object had something helpful to find the `route` method. Either method signature or perhaps the id of the corresponding `RouterFunction` bean. Seemed like getting the bean id into the JSON is doable. Therefore the JSON would look like:
```
          {
            "predicate": "((GET && /vets) && Accept: [text/plain])",
            "handler": "org.springframework.samples.petclinic.deviations.VetsRouter$$Lambda$868/443912630@369564bd",
            "details": {
              "handlerFunction": {
                "className": "org.springframework.samples.petclinic.deviations.VetsRouter$$Lambda$868/443912630"
              },
              "bean": "route"
            }
          },
```

The alternative is to look at the predicate. Matching the predicate string against source AST is complicated and unreliable since that string is a product of a `toString()` invocation. Doesn't seem right.

**Comment by wilkinsona** [comment]
Thanks for the suggestion. Unfortunately, I don't think it's possible to perform the `RouterFunction -> Bean` mapping as you have described.

Actuator is dealing with a `DispatcherHandler` which contains a single `RouterFunctionMapping`. It contains a single `RouterFunction` that isn't, itself, a bean. Typically, this composed `RouterFunction` is a composition of all of the application context's `RouterFunction` beans but this is an implementation detail that isn't exposed anywhere. We examine the composed function using a visitor. This provides us with access to each route (a request predicate and a handler function) but not to any information about the individual router functions.

In short, we would need some changes to be made in WebFlux.Fn to be able to support this. If you think this is worth pursuing, can you please open a Spring Framework issue and link to it here?

**Comment by bclozel** [comment]
Spring Framework recently added `RequestPredicate` visitors in [SPR-17322](https://jira.spring.io/browse/SPR-17322). I don't know if that's enough to enable that support, we need to take a look at it first.

**Comment by BoykoAlex** [comment]
Somehow I missed e-mail notification with your reply...
Indeed all rm functions are packed into a single one which actuator examines with a visitor...
Raised: https://jira.spring.io/browse/SPR-17579

**Comment by poutsma** [comment]
I left a comment on the corresponding JIRA that explains how to get to a `RouterFunction` bean id, see https://jira.spring.io/browse/SPR-17579?focusedCommentId=164021&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-164021

**Comment by BoykoAlex** [comment]
@wilkinsona does Arjens suggestion help in getting to the bean id on the actuator side?

**Comment by wilkinsona** [comment]
Yes, I think it does.
